### PR TITLE
auto update script

### PR DESCRIPTION
**Pull request by ian4hu**
Add auto update script

### CODE REVIEW

File: scripts/updates/auto-updates.sh [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	termux_version=$(echo "$termux_version" | cut -d: -f2-)
+	if [ "$termux_version" == "$termux_epoch" ]; then
+		# No epoch set.
+		termux_epoch=""
+	else
+		termux_epoch+=":"
+	fi
+
+	# Transform an actual version to regexp
+	version_regexp=$(echo "${termux_version}" | sed s@[0-9]@\\\\d+@g)
```
Comment (xeffyr) [inline scripts/updates/auto-updates.sh:47]: Are you sure you have read the "github-projects.txt" file and understand the format and purpose of fields?

This regexp is used to convert Git tags into format acceptable by `TERMUX_PKG_VERSION` and you can't recover the tag filtering regexp from the version.

Examples:
* `v1.0.0` - no regexp needed, just stripping the leading `v` is sufficient.
* `prog-1_25_1` - need regexp to extract and convert `1_25_1` into `1.25.1`.
  * https://github.com/cronie-crond/cronie/releases/tag/cronie-1.5.7: version is normalized, but still has prefix
  * https://github.com/curl/curl/releases/tag/curl-7_79_1: prefixed and numbers are separated by underscores.

File: scripts/updates/auto-updates.sh [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+: "${GIT_COMMIT_PACKAGES:=false}"
+# Push changes to remote.
+: "${GIT_PUSH_PACKAGES:=false}"
+
+if [ -z "${GITHUB_API_TOKEN-}" ]; then
+	echo "You need a Github Personal Access Token be set in variable GITHUB_API_TOKEN."
+	exit 1
+fi
+
+for package in $(ls -1 "${PKGSDIR}"); do
```
Comment (xeffyr) [inline scripts/updates/auto-updates.sh:19]: Never use `ls` in scripts. As for alternative, use
```
find $dir $options | while read -r line
```

File: scripts/updates/auto-updates.sh [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	unset auto_update_flag project version_regexp termux_version termux_epoch latest_version srcurl
+
+	# Extract some build vars
+	build_vars=$(set +e +u; . "${PKGSDIR}/${package}/build.sh" 2>/dev/null; \
+		echo "auto_update_flag=${TERMUX_PKG_AUTO_UPDATE}"; \
```
Comment (xeffyr) [inline scripts/updates/auto-updates.sh:24]: Slashes shouldn't be necessary there.

File: scripts/updates/auto-updates.sh [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		echo "termux_version=${TERMUX_PKG_VERSION}"; \
+		echo "srcurl=${TERMUX_PKG_SRCURL}" ;\
+	)
+
+	# Eval build vars to local variables
+	source <(echo "$build_vars")
+
+	if [ "${auto_update_flag}" != "true" ]; then
+		echo "[SKIP] ${package} is not auto updatable."
```
Comment (xeffyr) [inline scripts/updates/auto-updates.sh:33]: If package doesn't support auto updates, omit it. Don't print any useless messages. Remember that CI logs are subject for reviewing and any extra useless information makes it harder.

File: scripts/updates/auto-updates.sh [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	fi
+
+	# Transform an actual version to regexp
+	version_regexp=$(echo "${termux_version}" | sed s@[0-9]@\\\\d+@g)
+
+	# Extract github project from TERMUX_PKG_SRCURL
+	project="$(echo "${srcurl}" | grep github.com | cut -d / -f4-5)"
+	if [ -z "${project}" ]; then
+		echo "[SKIP] ${package}'s TERMUX_PKG_SRCURL is not a github archive url."
+		continue
```
Comment (xeffyr) [inline scripts/updates/auto-updates.sh:53]: Here must be an error. If package was configured for auto updates by mistake, we should get notification that auto updates have a problem.

File: scripts/updates/auto-updates.sh [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	latest_version=${latest_version#v}
+
+	# If needed, filter version numbers from tag by using regexp.
+	if [ -n "$version_regexp" ]; then
+		latest_version=$(grep -oP "$version_regexp" <<< "$latest_version" || true)
+	fi
+
+	if [[ -z "$latest_version" || "$latest_version" = "null" ]]; then
+		echo "[SKIP] Failed to get latest version for '${package}'."
+		continue
```
Comment (xeffyr) [inline scripts/updates/auto-updates.sh:69]: This must return error too.

**Review by xeffyr — CHANGES_REQUESTED** [review]
Few more issues. Also by the way why you are submitting changes as new file and why it contains some more modifications than just in-`build.sh` auto updates config?

As for now, that's cannot be accepted.

Summary:
1. Find the way to define regexp to extract the version number from Git tag returned by GitHub API. Otherwise if you want to omit the definition of regexp, find the way to get correct version number directly (i.e. `x.y.z` and not `prog-x.y.z` or `x_y_z`).
2. Fix scripting issues, e.g. use of `ls` for getting directory listing.
3. Remove unnecessary changes, especially absent in original updater script.
4. Remember that script must fail on the first unexpected case, so we will get a notification about failed GitHub Actions run.

File: scripts/updates/auto-updates.sh (hunk 1 above)
Comment (ian4hu) [reply scripts/updates/auto-updates.sh:47]: Thanks for your reviewing. the `version_regexp` needs more work, I will try it.

**Comment by ian4hu** [comment]
@xeffyr  Could we put the lines in `github-projects.txt` as the `TERMUX_PKG_AUTO_UPDATE` like

```shell
TERMUX_PKG_AUTO_UPDATE=true or (version_regexp)

# Enable auto update and use default version format v1.2.3
TERMUX_PKG_AUTO_UPDATE=true

TERMUX_PKG_AUTO_UPDATE="\d+.\d+.\d+"

```

**Comment by ian4hu** [comment]
#7621 We could discuss solution here

**Comment by xeffyr** [comment]
This will be sufficient to solve the tag filtering issue:
```
TERMUX_PKG_AUTO_UPDATE=true
TERMUX_PKG_AUTO_UPDATE_TAG_REGEXP="\\d+\.\\d+\.\\d+"
```

**Comment by ian4hu** [comment]
@xeffyr Would you prefer get the new auto-update in the `check-updates.sh` ? or create a new one?

**Comment by xeffyr** [comment]
Make changes to check-updates.sh

File: scripts/updates/check-updates.sh [hunk 7]
```diff
@@ -10,85 +10,115 @@ BASEDIR=$(dirname "$(realpath "$0")")
...
-					echo "Failed to update checksum for '${package}', skipping..."
-					git checkout -- "${BASEDIR}/../../packages/${package}"
+	# Translate "_" into ".".
+	latest_version=${latest_version//_/.}
+
+	latest_version_raw=$latest_version
+	# If needed, filter version numbers from tag by using regexp.
+	if [ -n "$version_regexp" ]; then
+		latest_version=$(grep -oP "$version_regexp" <<< "$latest_version" || true)
+	fi
```
Comment (xeffyr) [inline scripts/updates/check-updates.sh:79]: You have messed up the order by which version tags are processed. All translations except stripping the leading `v` should be done after regexp filter.

Anyway, I'm fixing script on my own now as well as test it.

**Comment by xeffyr** [comment]
Everything is available on `master` now, thanks.

**Comment by ian4hu** [comment]
@xeffyr Another suggestion: Some package can be failed by some reasons

- Build failed: compiler error/ missing new deps etc
image

- Patch failed: patch can not apply to new version source tree.
image

We should not fail all packages auto-update by few packages auto-update failure. Should we have a failure tolerance in auto-update?

**Pull request closed without merging**